spec/std/big/big_decimal_spec.cr

```diff
@@ -0,0 +1,307 @@
+    (BigDecimal.new(1) < 1).should be_false
+    (2 < BigDecimal.new(1)).should be_false
+    (BigDecimal.new(2) < 1).should be_false
+    (BigDecimal.new("-1") > BigDecimal.new("1")).should be_false
+
+    (1 > BigDecimal.new(1)).should be_false
+    (BigDecimal.new(1) > 1).should be_false
+    (2 > BigDecimal.new(1)).should be_true
+    (BigDecimal.new(2) > 1).should be_true
+    (BigDecimal.new("-1") < BigDecimal.new("1")).should be_true
+
+    (1 >= BigDecimal.new(1)).should be_true
+    (2 >= BigDecimal.new(1)).should be_true
+
+    (1 <= BigDecimal.new(1)).should be_true
+    (0 <= BigDecimal.new(1)).should be_true
+
+    (BigDecimal.new("6.5") > 7).should be_false
+    (BigDecimal.new("7.5") > 6).should be_true
+  end
+
+  it "keeps precision" do
+    one_thousandth = BigDecimal.new("0.001")
+    one = BigDecimal.new("1")
+
+    x = BigDecimal.new
+    1000.times do
+      x += one_thousandth
+    end
+    one.should eq(x)
+
+    x = BigDecimal.new("2")
+    1000.times do
+      x -= one_thousandth
+    end
+    one.should eq(x)
+  end
+
+  it "converts to string" do
+    BigDecimal.new.to_s.should eq "0"
+    BigDecimal.new(0).to_s.should eq "0"
+    BigDecimal.new(1).to_s.should eq "1"
+    BigDecimal.new(-1).to_s.should eq "-1"
+    BigDecimal.new("0.01").to_s.should eq "0.01"
+    BigDecimal.new("1.0").to_s.should eq "1"
+    BigDecimal.new("-1.0").to_s.should eq "-1"
+    BigDecimal.new("1.000").to_s.should eq "1"
+    BigDecimal.new("-1.000").to_s.should eq "-1"
+    BigDecimal.new("1.0001").to_s.should eq "1.0001"
+    BigDecimal.new("-1.0001").to_s.should eq "-1.0001"
+
+    (BigDecimal.new(1).div(BigDecimal.new(3), 9)).to_s.should eq "0.333333333"
+    (BigDecimal.new(1000).div(BigDecimal.new(3000), 9)).to_s.should eq "0.333333333"
+    (BigDecimal.new(1).div(BigDecimal.new(3000), 9)).to_s.should eq "0.000333333"
+
+    (BigDecimal.new("112839719283").div(BigDecimal.new("3123779"), 9)).to_s.should eq "36122.824080384"
+    (BigDecimal.new("112839719283").div(BigDecimal.new("3123779"), 14)).to_s.should eq "36122.8240803846879"
+
+    BigDecimal.new(1, 2).to_s.should eq "0.01"
+    BigDecimal.new(100, 4).to_s.should eq "0.01"
+  end
+
+  it "hashes" do
+    bd1 = BigDecimal.new("123.456")
+    bd2 = BigDecimal.new("0.12345")
+    bd3 = BigDecimal.new("1.23456")
+    bd4 = BigDecimal.new("123456")
+    bd5 = BigDecimal.new("0")
+
+    hash = {} of BigDecimal => String
+    hash[bd1] = "bd1"
+    hash[bd2] = "bd2"
+    hash[bd3] = "bd3"
+    hash[bd4] = "bd4"
+    hash[bd5] = "bd5"
+
+    # regular cases
+    hash[BigDecimal.new("123.456")].should eq "bd1"
+    hash[BigDecimal.new("0.12345")].should eq "bd2"
+    hash[BigDecimal.new("1.23456")].should eq "bd3"
+    hash[BigDecimal.new("123456")].should eq "bd4"
+    hash[BigDecimal.new("0")].should eq "bd5"
+
+    # not found
+    expect_raises(KeyError) do
+      hash[BigDecimal.new("4")]
+    end
+  end
+
+  it "upkeeps hashing contract" do
+    # a == b <=> h[a] == h[b]
```

Comment:
This test does actually check for `a = b ⇔ h[a] = h[b]` but that condition is impossible to satisfy because of pigeonhole principle.
The invariant is `a = b ⇒ h[a] = h[b]`. Just check `BigDecimal.new(1, 2).hash == BigDecimal.new(100, 4).hash` and be done with it.

Reply:
Spec changed as you recommended.